File: src/main/java/rars/settings/SettingsThemePresets.java
package rars.settings;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Unmodifiable;
import rars.venus.editors.TokenStyle;

import java.awt.*;
import java.util.List;
import java.util.Map;
import java.util.function.UnaryOperator;

import static java.util.Map.entry;

public final class SettingsThemePresets {
    public static final @NotNull SettingsTheme LIGHT_THEME = buildTheme(theme -> {
        theme.foregroundColor = Color.BLACK;
        theme.backgroundColor = Color.WHITE;
        theme.lineHighlightColor = new Color(242, 242, 242);
        theme.selectionColor = new Color(229, 235, 235);
        theme.caretColor = Color.DARK_GRAY;

        final var errorColor = new Color(255, 0, 0);
        final var commentColor = new Color(0, 128, 0);
        final var directiveColor = new Color(0, 134, 116);
        final var registerColor = new Color(169, 103, 0);
        final var numberColor = new Color(100, 0, 200);
        final var stringColor = new Color(175, 1, 121);
        final var labelColor = new Color(128, 128, 128);
        final var instructionColor = new Color(33, 33, 161);
        final var macroParameterColor = new Color(178, 49, 27);

        theme.tokenStyles.putAll(Map.ofEntries(
            entry(TokenSettingKey.ERROR, TokenStyle.underline(errorColor)),
            entry(TokenSettingKey.COMMENT, TokenStyle.italic(commentColor)),
            entry(TokenSettingKey.DIRECTIVE, TokenStyle.bold(directiveColor)),

            entry(TokenSettingKey.REGISTER_NAME, TokenStyle.bold(registerColor)),

            entry(TokenSettingKey.NUMBER, TokenStyle.bold(numberColor)),
            entry(TokenSettingKey.STRING, TokenStyle.bold(stringColor)),

            entry(TokenSettingKey.LABEL, TokenStyle.bold(labelColor)),
            entry(TokenSettingKey.INSTRUCTION, TokenStyle.bold(instructionColor)),

            entry(TokenSettingKey.MACRO_PARAMETER, TokenStyle.bold(macroParameterColor))
        ));
        return theme;
    });
    public static final @NotNull SettingsTheme DARK_THEME = buildTheme(theme -> {
        theme.foregroundColor = Color.WHITE;
        theme.backgroundColor = new Color(31, 31, 31);
        theme.lineHighlightColor = new Color(47, 47, 47);
        theme.selectionColor = new Color(61, 61, 61);
        theme.caretColor = Color.LIGHT_GRAY;

        final var errorColor = new Color(255, 0, 0);
        final var commentColor = new Color(0, 159, 0);
        final var directiveColor = new Color(0, 183, 159);
        final var registerColor = new Color(222, 134, 0);
        final var numberColor = new Color(159, 57, 255);
        final var stringColor = new Color(255, 0, 176);
        final var labelColor = new Color(161, 161, 161);
        final var instructionColor = new Color(80, 80, 253);
        final var macroParameterColor = new Color(255, 70, 38);

        theme.tokenStyles.putAll(Map.ofEntries(
            entry(TokenSettingKey.ERROR, TokenStyle.underline(errorColor)),
            entry(TokenSettingKey.COMMENT, TokenStyle.italic(commentColor)),
            entry(TokenSettingKey.DIRECTIVE, TokenStyle.bold(directiveColor)),

            entry(TokenSettingKey.REGISTER_NAME, TokenStyle.bold(registerColor)),

            entry(TokenSettingKey.NUMBER, TokenStyle.bold(numberColor)),
            entry(TokenSettingKey.STRING, TokenStyle.bold(stringColor)),

            entry(TokenSettingKey.LABEL, TokenStyle.bold(labelColor)),
            entry(TokenSettingKey.INSTRUCTION, TokenStyle.bold(instructionColor)),

            entry(TokenSettingKey.MACRO_PARAMETER, TokenStyle.bold(macroParameterColor))
        ));
        return theme;
    });
    public static final @NotNull
    @Unmodifiable List<@NotNull ThemeEntry> THEMES = List.of(
        new ThemeEntry("Default light", LIGHT_THEME),
        new ThemeEntry("Default dark", DARK_THEME)
    );

    private static @NotNull SettingsTheme buildTheme(final @NotNull UnaryOperator<SettingsTheme> builderFunction) {
        return builderFunction.apply(new SettingsTheme(
            Color.WHITE,
            Color.WHITE,
            Color.WHITE,
            Color.WHITE,
            Color.WHITE,
            Map.of()
        ));
    }

    public record ThemeEntry(@NotNull String name, @NotNull SettingsTheme theme) {
    }
}

File: src/main/java/rars/settings/SettingsThemePresets.kt
@file:Suppress("DuplicatedCode")

package rars.settings

import rars.venus.editors.TokenStyle
import java.awt.Color

object SettingsThemePresets {
    val LIGHT_THEME: SettingsTheme by lazyTheme {
        foregroundColor = Color.BLACK
        backgroundColor = Color.WHITE
        lineHighlightColor = Color(242, 242, 242)
        selectionColor = Color(229, 235, 235)
        caretColor = Color.DARK_GRAY

        val errorColor = Color(255, 0, 0)
        val commentColor = Color(0, 128, 0)
        val directiveColor = Color(0, 134, 116)
        val registerColor = Color(169, 103, 0)
        val numberColor = Color(100, 0, 200)
        val stringColor = Color(175, 1, 121)
        val labelColor = Color(128, 128, 128)
        val instructionColor = Color(33, 33, 161)
        val macroParameterColor = Color(178, 49, 27)

        tokenStyles[TokenSettingKey.ERROR] = TokenStyle.underline(errorColor)
        tokenStyles[TokenSettingKey.COMMENT] = TokenStyle.italic(commentColor)
        tokenStyles[TokenSettingKey.DIRECTIVE] = TokenStyle.bold(directiveColor)

        tokenStyles[TokenSettingKey.REGISTER_NAME] = TokenStyle.bold(registerColor)

        tokenStyles[TokenSettingKey.NUMBER] = TokenStyle.bold(numberColor)
        tokenStyles[TokenSettingKey.STRING] = TokenStyle.bold(stringColor)

        tokenStyles[TokenSettingKey.LABEL] = TokenStyle.bold(labelColor)
        tokenStyles[TokenSettingKey.INSTRUCTION] = TokenStyle.bold(instructionColor)

        tokenStyles[TokenSettingKey.MACRO_PARAMETER] = TokenStyle.bold(macroParameterColor)
    }
    val DARK_THEME: SettingsTheme by lazyTheme {
        foregroundColor = Color.WHITE
        backgroundColor = Color(31, 31, 31)
        lineHighlightColor = Color(47, 47, 47)
        selectionColor = Color(61, 61, 61)
        caretColor = Color.LIGHT_GRAY

        val errorColor = Color(255, 0, 0)
        val commentColor = Color(0, 159, 0)
        val directiveColor = Color(0, 183, 159)
        val registerColor = Color(222, 134, 0)
        val numberColor = Color(159, 57, 255)
        val stringColor = Color(255, 0, 176)
        val labelColor = Color(161, 161, 161)
        val instructionColor = Color(80, 80, 253)
        val macroParameterColor = Color(255, 70, 38)

        tokenStyles[TokenSettingKey.ERROR] = TokenStyle.underline(errorColor)
        tokenStyles[TokenSettingKey.COMMENT] = TokenStyle.italic(commentColor)
        tokenStyles[TokenSettingKey.DIRECTIVE] = TokenStyle.bold(directiveColor)

        tokenStyles[TokenSettingKey.REGISTER_NAME] = TokenStyle.bold(registerColor)

        tokenStyles[TokenSettingKey.NUMBER] = TokenStyle.bold(numberColor)
        tokenStyles[TokenSettingKey.STRING] = TokenStyle.bold(stringColor)

        tokenStyles[TokenSettingKey.LABEL] = TokenStyle.bold(labelColor)
        tokenStyles[TokenSettingKey.INSTRUCTION] = TokenStyle.bold(instructionColor)

        tokenStyles[TokenSettingKey.MACRO_PARAMETER] = TokenStyle.bold(macroParameterColor)
    }

    @JvmStatic
    val THEMES: List<ThemeEntry> by lazy {
        listOf(
            ThemeEntry("Default light", LIGHT_THEME),
            ThemeEntry("Default dark", DARK_THEME)
        )
    }

    private fun lazyTheme(builderFunction: SettingsTheme.() -> Unit): Lazy<SettingsTheme> = lazy {
        buildTheme(builderFunction)
    }

    private fun buildTheme(builderFunction: SettingsTheme.() -> Unit): SettingsTheme = SettingsTheme(
        Color.WHITE,
        Color.WHITE,
        Color.WHITE,
        Color.WHITE,
        Color.WHITE,
        mapOf<TokenSettingKey, TokenStyle>()
    ).apply(builderFunction)

    data class ThemeEntry(val name: String, val theme: SettingsTheme)
}
